Add new  workflow `DiscoverGranulesBackground`

flamingbear opened the pull request:
**Summary**

This discovers and writes granules to a throttled background queue.  This
allows discovery and ingest of larger numbers of granules without running into
limits with lambda concurrency.

Exposes necessary variables to enable throttled queue.
addresses [CUMULUS-2630: CORS errors in Dashboard](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2630)

## Changes

## PR Checklist

- [ X ] Update CHANGELOG
- [ ] Unit tests
- [ X ] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

flamingbear commented:
I have questions as to whether this is how I should do it. So if this is extremely wrong, feel free to say that.

Jkovarik commented on CHANGELOG.md:
```diff
@@ -18,6 +18,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - **CUMULUS-2583**
   - Adds `queued` as option for granule's `status` field
+- **CUMULUS-2630**
+  - Adds a new workflow `DiscoverGranulesBackground` that discovers and writes
```
Nit: to example development project

What's the purpose of adding this?   To facilitate dev testing, or for users?  If the latter, should we have a PR from 2630 to the template/docs/???

flamingbear replied:
I answered in reverse order.  It's for dev testing.  But IDK if it would be useful for users.

markdboyd replied:
For users, this is the documentation on how to use the feature in question:

https://nasa.github.io/cumulus/docs/data-cookbooks/throttling-queued-executions

We don't have a full example workflow in our template repo for it or anything though. Whether we should is a fair question

Jkovarik commented on tf-modules/cumulus/README.md:
```diff
@@ -24,6 +24,7 @@ This module's outputs are listed in [ouputs.tf](./outputs.tf). Notable values th
 - **report_executions_sns_topic_arn** - The ARN of the SNS topic used for reporting the status of Cumulus workflow executions. You will need this ARN if you want to publish to this topic directly.
 - **report_granules_sns_topic_arn** - The ARN of the SNS topic used for reporting the ingest status of granules for Cumulus workflows. You will need this ARN if you want to publish to this topic directly.
 - **report_pdrs_sns_topic_arn** - The ARN of the SNS topic used for reporting the ingest status of PDRs for Cumulus workflows. You will need this ARN if you want to publish to this topic directly.
+- **background_queue_url** - The url of the background queue.
```
This seems .... terse.    Can we be a bit more descriptive here?

flamingbear replied:
Yeah it's terse. I'll remove for `outputs.tf`

flamingbear replied:
67a94876f3

Jkovarik commented on tf-modules/cumulus/outputs.tf:
```diff
@@ -169,3 +169,7 @@ output "ecs_cluster_name" {
   value = module.ingest.start_sf_queue_url
 }
+
+output "background_queue_url" {
```
So this is the functional change..... I'm not sure how this ties to CUMULUS-2630?

flamingbear replied:
So CUMULUS-2630 is closed, but I needed this work to generate lots of data for our dashboards.  I had this in my stashes and Katie has asked for the tools I used to make lots of data, so I thought I would put it in here as an example of a throttled workflow for discovering lots of granules.  If it's not appropriate I'm fine leaving it out.

markdboyd replied:
So I'm a bit hesitant to expose this. It's an internal queue that is throttled, but users can't control the throttling limit:

https://github.com/nasa/cumulus/blob/master/tf-modules/ingest/message_template.tf#L47

It exists solely for the purposes of us having a throttled queue built into Cumulus, which we send reingest operations to by default because we don't care about them happening very quickly: 

https://github.com/nasa/cumulus/blob/master/tf-modules/ingest/message_template.tf#L47

I'm inclined to say that for anyone (including us for testing) who wants to have their own throttled queue setup, they should implement a new queue so that they can control the execution limit:

https://nasa.github.io/cumulus/docs/data-cookbooks/throttling-queued-executions

Beyond that, I'm okay with having a  discover-granules workflow deployed with our example stacks that supports throttling by default.

flamingbear replied:
Ok, this goes along with the, if this is all wrong tell me.  I can put this back as private and create another queue in our example

Jkovarik reviewed (changes requested):
Just a background question and a couple of nits, the implementation looks sound.

flamingbear commented:
@markdboyd I removed the dependency on the background_job queue and added a new queue to the example/cumulus-tf project.  This is ready for re-review.

markdboyd commented on example/cumulus-tf/main.tf:
```diff
@@ -177,7 +181,7 @@ module "cumulus" {
   log_destination_arn = var.log_destination_arn
 
-  # Cumulus Distribution settings. Remove/comment if not using Cumulus Distribution.
+  # Cumulus Distribution settings. Remove/comment if not using Cumulus Distribution. (are using TEA)
```
What does `(are using TEA)` mean?

flamingbear replied:
I means I couldn't parse the different between using cumulus distribution vs using TEA.
So not using Cumulus Distribution == using TEA.  Right above this we say "Thin Egress App settings. Uncomment to use TEA."
then here we say remove comment if not using Cumulus Distribution.  So now that I've explained it, what about? 
```suggestion
  # Cumulus Distribution settings. Remove/comment to use TEA```

markdboyd replied:
```suggestion
  # Cumulus Distribution settings. Remove/comment to use TEA
```

You had some extra backticks in your suggestion. Otherwise, looks good

markdboyd commented on CHANGELOG.md:
```diff
@@ -43,6 +43,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
   - Adds helper `associateExecutionWithGranule` to `@cumulus/api-client/granules`
 - **CUMULUS-2583**
   - Adds `queued` as option for granule's `status` field
+- **CUMULUS-2630**
```
This changelog note shouldn't be nested under 9.6.0

flamingbear replied:
833f3cd6b7

markdboyd commented:
@flamingbear Great work! Just one more change for the changelog
